src/pool/srv_pool_scrub_ult.c

```diff
@@ -0,0 +1,367 @@
+	struct scrub_ctx	 ctx = {0};
+	struct ds_pool_child	*child = arg;
+	struct dss_module_info	*dmi = dss_get_module_info();
+	uuid_t			 pool_uuid;
+	daos_handle_t		 poh;
+	int			 tgt_id;
+	int			 rc;
+
+	poh = child->spc_hdl;
+	uuid_copy(pool_uuid, child->spc_uuid);
+	tgt_id = dmi->dmi_tgt_id;
+
+	C_TRACE(DF_PTGT": Scrubbing ULT started\n", DP_PTGT(pool_uuid, tgt_id));
+
+	if (child->spc_scrubbing_req == NULL)
+		return;
+
+	uuid_copy(ctx.sc_pool_uuid, pool_uuid);
+	ctx.sc_vos_pool_hdl = poh;
+	ctx.sc_sleep_fn = sleep_fn;
+	ctx.sc_yield_fn = yield_fn;
+	ctx.sc_sched_arg = child->spc_scrubbing_req;
+	ctx.sc_pool = child->spc_pool;
+	ctx.sc_cont_lookup_fn = cont_lookup_cb;
+	ctx.sc_cont_put_fn = cont_put_cb;
+	ctx.sc_cont_is_stopping_fn = cont_is_stopping_cb;
+	ctx.sc_dmi =  dss_get_module_info();
+	ctx.sc_drain_pool_tgt_fn = drain_pool_tgt_cb;
+
+	sc_add_pool_metrics(&ctx);
+	while (!dss_ult_exiting(child->spc_scrubbing_req)) {
+		rc = vos_scrub_pool(&ctx);
+		if (rc != 0)
+			break;
+	}
+}
+
+int
+ds_start_scrubbing_ult(struct ds_pool_child *child)
+{
+	struct dss_module_info	*dmi = dss_get_module_info();
+	struct sched_req_attr	 attr;
+
+	D_ASSERT(child != NULL);
+	D_ASSERT(child->spc_scrubbing_req == NULL);
+
+	/** Don't even create the ULT if scrubbing is disabled. */
+	if (!scrubbing_is_enabled()) {
+		C_TRACE("Checksum scrubbing DISABLED. "
+			"xs_id: %d, tgt_id: %d, ctx_id: %d, ",
+			dmi->dmi_xs_id, dmi->dmi_tgt_id, dmi->dmi_ctx_id);
+		return 0;
+	}
+
+	C_TRACE("Checksum scrubbing ENABLED. "
+		"xs_id: %d, tgt_id: %d, ctx_id: %d, ",
+		dmi->dmi_xs_id, dmi->dmi_tgt_id, dmi->dmi_ctx_id);
+
+	/* There will be several levels iteration, such as pool, container, object, and lower,
+	 * and so on. Let's use DSS_DEEP_STACK_SZ to avoid ULT overflow.
+	 */
+	sched_req_attr_init(&attr, SCHED_REQ_SCRUB, &child->spc_uuid);
+	child->spc_scrubbing_req = sched_create_ult(&attr, scrubbing_ult, child,
+						    DSS_DEEP_STACK_SZ);
+	if (child->spc_scrubbing_req == NULL) {
+		D_ERROR(DF_UUID"[%d]: Failed to create Scrubbing ULT.n",
+			DP_UUID(child->spc_uuid), dmi->dmi_tgt_id);
+		return -DER_NOMEM;
+	}
+
+	return 0;
+}
+//
+///** Setup and create the scrubbing ult */
+//int
+//ds_start_scrubbing_ult(struct ds_pool_child *child)
+//{
+//	struct dss_module_info	*dmi = dss_get_module_info();
+//	struct sched_req_attr	 attr = {0};
+//	ABT_thread		 thread = ABT_THREAD_NULL;
+//	int			 rc;
+//
+//	D_ASSERT(child != NULL);
+//	D_ASSERT(child->spc_scrubbing_req == NULL);
+//
+//	/** Don't even create the ULT if scrubbing is disabled. */
+//	if (!scrubbing_is_enabled()) {
+//		C_TRACE(DF_PTGT": Checksum scrubbing DISABLED.\n",
+//			DP_PTGT(child->spc_uuid, dmi->dmi_tgt_id));
+//		return 0;
+//	}
+//
+//	C_TRACE(DF_PTGT": Checksum scrubbing Enabled. Creating ULT.\n",
+//		DP_PTGT(child->spc_uuid, dmi->dmi_tgt_id));
+//
+//	/* There will be several levels iteration, such as pool, container,
+//	 * object, and lower, and so on. Let's use DSS_DEEP_STACK_SZ to avoid
```

Comment:
(style)  Block comments use * on subsequent lines
(style)  do not use C99 // comments